stoppable_workers.go

```diff
@@ -0,0 +1,88 @@
+package utils
+
+import (
+	"context"
+	"sync"
+)
+
+// StoppableWorkers is a collection of goroutines that can be stopped at a
+// later time.
+type StoppableWorkers struct {
+	// Use a `sync.RWMutex` isntead of a `sync.Mutex` so that additions of new
+	// workers do not lock with each other in any way. We want
+	// as-fast-as-possible worker addition.
+	mu         sync.RWMutex
+	ctx        context.Context
+	cancelFunc func()
+
+	workers sync.WaitGroup
+}
+
+// NewStoppableWorkers creates a new StoppableWorkers instance. The instance's
+// context will be derived from passed in context.
+func NewStoppableWorkers(ctx context.Context) *StoppableWorkers {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	return &StoppableWorkers{ctx: ctx, cancelFunc: cancelFunc}
+}
+
+// NewBackgroundStoppableWorkers creates a new StoppableWorkers instance. The
+// instance's context will be derived from `context.Background()`. The passed
+// in workers will be `Add`ed. Workers:
+//
+//   - MUST respond appropriately to errors on the context parameter.
+//   - MUST NOT add more workers to the `StoppableWorkers` group to which
+//     they belong.
```

Comment:
Why not? I don't see any obvious problems with a worker spawning another worker. There's no deadlock in here: the first worker's goroutine would block until the StoppableWorkers' mutex was unlocked, then spawn the new goroutine afterward.

Reply:
What if:
1. A worker is `Add`ed
2. `Stop` is called
      a. Mutex is `Lock`ed
      b. Context is canceled
      b. `Wait` is called
3. The worker adds another subworker _before it has a chance to check its context_
5. That second worker gets caught on `RLock`ing the locked mutex
6. The first worker never marks itself as `Done()` as the worker it spawns gets caught on `RLock`
7. `Wait` hangs forever

I realize this may seem like a farfetched scenario, but NetCode has encountered it in our code, and I have found it dangerous to have "subworkers" accessing the same concurrency primitives that their parents do. We can avoid complex situations like the one above by ensuring that workers with subworkers have their _own_ `StoppableWorkers` instance.

Reply:
Ah, very good insight! Thanks for explaining.

Reply:
No prob 😎 . Thanks for writing all the original code here, too.